### PR TITLE
[RISCV] Add FPR16 regbank and start legalizing f16 operations for Zfh.

### PR DESCRIPTION
**Pull request by topperc**

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:379]: Doesn't it work?
```
static LegalityPredicate typeIsScalarFPArith(unsigned TypeIdx,
                                             const RISCVSubtarget &ST) {
  return [=, &ST](const LegalityQuery &Query) {
    return Query.Types[TypeIdx].isScalar() &&
           ((ST.hasStdExtF() && Query.Types[TypeIdx].getSizeInBits() == 32) ||
            (ST.hasStdExtD() && Query.Types[TypeIdx].getSizeInBits() == 64) ||
            (ST.hasStdExtZfh() && Query.Types[TypeIdx].getSizeInBits() == 16));
  };
}
```

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir:193]: Missing tests for G_FMAXNUM/G_FMINNUM/G_COPYSIGN.

**Review by dtcxzyw — COMMENTED** [review]
Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.

Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:379]: It should, but it requires many more tests to cover every other call to typeIsScalarFPArith

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir
Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/fp-arith-f16.mir:193]: Oops. I naively assumed that all the f32 tests were together when I copied this over.

**Comment by topperc** [comment]
> Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.

Why is that blocking?

> Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?

Yes.

**Comment by dtcxzyw** [comment]
> > Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.
> 
> Why is that blocking?

As we widen the type from f16 to f32, we need to implement G_FPTRUNC/G_FPEXT/G_ANYEXT for s16<->s32.

> 
> > Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?
> 
> Yes.

Great! TBH I don't have enough bandwidth to work on this. I will post my existing code later. Hope this helps you.

**Comment by dtcxzyw** [comment]
> > > Oops, I already have an unsubmitted patch for half fp arith :) It was blocked by supporting half fp with only F or F + Zfhmin.
> > 
> > 
> > Why is that blocking?
> 
> As we widen the type from f16 to f32, we need to implement G_FPTRUNC/G_FPEXT/G_ANYEXT for s16<->s32.
> 
> > > Do you plan to support other ops (fcmp/fconstant/fp<->fp/fp<->int)?
> > 
> > 
> > Yes.
> 
> Great! TBH I don't have enough bandwidth to work on this. I will post my existing code later. Hope this helps you.

See https://github.com/dtcxzyw/llvm-project/commit/7948041a4eaf9f89247c3f75b8afb2a5b998f642

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!
Wait for an additional approval from other reviewers.

**Review by arsenm — APPROVED** [review]

**Review by michaelmaitland — APPROVED** [review]
LGTM

**Pull request merged**
